### PR TITLE
eth2util/deposit: refactor deposit data and add tests

### PR DESCRIPTION
**Pull request by xenowits**
* refactor `deposit data` to use types from `eth2p0` ([ex](https://pkg.go.dev/github.com/attestantio/go-eth2-client/spec/phase0@v0.11.3#DepositData)).
* Test generated deposit data against a sample output generated using [eth2 cli](https://github.com/ethereum/staking-deposit-cli).

category: refactor
ticket: #481

### CODE REVIEW

File: eth2util/deposit/testdata/sd_cli_keystore_json.golden [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (corverroos) [inline eth2util/deposit/testdata/sd_cli_keystore_json.golden:1]: store the privatekey as in the test code directly:
```
const privkey = "acb123adc4212367"
```

File: eth2util/keystore/keystore.go [hunk 2]
```diff
@@ -108,8 +108,8 @@ func LoadKeys(dir string) ([]*bls_sig.SecretKey, error) {
 	return resp, nil
 }
 
-// keystore json file representation as a Go struct.
-type keystore struct {
+// Keystore json file representation as a Go struct.
+type Keystore struct {
```
Comment (corverroos) [inline eth2util/keystore/keystore.go:112]: Best not refactor other packages just for testing.

File: eth2util/deposit/deposit_data_test.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	"github.com/obolnetwork/charon/eth2util/keystore"
+	"github.com/obolnetwork/charon/tbls/tblsconv"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"os"
+	"path"
+	"testing"
+)
+
+func TestDepositData1Serialization(t *testing.T) {
```
Comment (corverroos) [inline eth2util/deposit/deposit_data_test.go:32]: As single test should be enough:

 ```
 root, err := deposit.MessageRoot(...)
 sig := testSign(root)
 resp, err := deposit.MarshalDepositData(..., sig)
 testutil.RequireGoldenBytes(t, resp)
 ```

File: eth2util/keystore/keystore.go (hunk 2 above)
Comment (leolara) [reply eth2util/keystore/keystore.go:112]: it is better code and it is already done and very short. In general I agree of not getting into other stuff outside the PR goal

**Comment by leolara** [comment]
@xenowits I don't undestand why there is a feature flag called eth2util? is it eth2util a feature? the name does not suggest a feature

**Comment by corverroos** [comment]
Please rebase this on latest main 🙏

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 repos:
   # First run code formatters
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v4.1.0
+    rev: v4.2.0
```
Comment (corverroos) [inline .pre-commit-config.yaml:4]: please move this a separate PR.

File: eth2util/deposit/deposit_data.go [hunk 5]
```diff
@@ -18,142 +18,219 @@ package deposit
...
+	zeroBytes32 = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
 
-	// DepositMessageRoot is the hash tree root of DepositMessage.
-	DepositMessageRoot eth2p0.Root
+	// DOMAIN_DEPOSIT. See spec: https://benjaminion.xyz/eth2-annotated-spec/phase0/beacon-chain/#domain-types
+	depositDomainType = []byte{0x03, 0x00, 0x00, 0x00}
+)
+
+// GetMessageRoot returns both the hash root and the signing root of the deposit message.
+func GetMessageRoot(pubkey eth2p0.BLSPubKey, withdrawalCreds [32]byte, forkVersion eth2p0.Version) (eth2p0.Root, eth2p0.Root, error) {
```
Comment (corverroos) [inline eth2util/deposit/deposit_data.go:49]: when returning identical types, it is best to use named parameters so it is clear which one is which.

Comment (corverroos) [reply eth2util/deposit/deposit_data.go:49]: Wouldn't a `withdrawalAddr string` be a more convenient format for users of this package?

File: eth2util/deposit/deposit_data.go [hunk 6]
```diff
@@ -18,142 +18,219 @@ package deposit
...
+	}
+
+	return hashRoot, signingRoot, nil
 }
 
-func (d depositData) HashTreeRootWith(hh *ssz.Hasher) error {
-	idx := hh.Index()
+// NewDepositData returns the json serialized DepositData.
+func NewDepositData(pubkey eth2p0.BLSPubKey, withdrawalAddr string, sig eth2p0.BLSSignature, network string) ([]byte, error) {
+	forkVersion := NetworkToForkVersion(network)
```
Comment (corverroos) [inline eth2util/deposit/deposit_data.go:94]: I would unexport all functions, except `MarshalDepositData` and `GetMessageRoot`

**Comment by xenowits** [comment]
> @xenowits I don't undestand why there is a feature flag called eth2util? is it eth2util a feature? the name does not suggest a feature

I removed the feature_flag as it was not necessary.

File: eth2util/deposit/deposit_data.go [hunk 7]
```diff
@@ -18,142 +18,219 @@ package deposit
...
 
-	// Eth1WithdrawalAddress is the Ethereum withdrawal address.
-	Eth1WithdrawalAddress string
+	// zeroes refers to a zeroed out 32 byte array.
+	zeroBytes32 = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
 
-	// DepositMessageRoot is the hash tree root of DepositMessage.
-	DepositMessageRoot eth2p0.Root
+	// DOMAIN_DEPOSIT. See spec: https://benjaminion.xyz/eth2-annotated-spec/phase0/beacon-chain/#domain-types
+	depositDomainType = []byte{0x03, 0x00, 0x00, 0x00}
```
Comment (corverroos) [inline eth2util/deposit/deposit_data.go:45]: wrap in proper type `= eth2p0.DomainType([4]byte{0x03, 0x00, 0x00, 0x00})`

Comment (corverroos) [reply eth2util/deposit/deposit_data.go:45]: same for zeroBytes32

File: eth2util/deposit/deposit_data.go [hunk 8]
```diff
@@ -18,142 +18,219 @@ package deposit
...
 
-	// Field 4 'Signature'
-	hh.PutBytes(d.Signature[:])
+	forkData := &eth2p0.ForkData{
+		CurrentVersion:        forkVersion,
+		GenesisValidatorsRoot: genesisValidatorsRoot,
+	}
+	root, err := forkData.HashTreeRoot()
+	if err != nil {
+		return eth2p0.Domain{}, errors.Wrap(err, "failed to calculate signature domain")
```
Comment (corverroos) [inline eth2util/deposit/deposit_data.go:142]: nit: can be more concise: `hash fork data`

File: eth2util/deposit/deposit_data_test.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+var (
+	depositDataFile = path.Join("testdata", "TestDepositData.golden")
+	privKey         = "07e0355752a16fdc473d01676f7f82594991ea830eea928d8e2254dfd98d4beb"
+	withdrawalAddr  = "0xc0404ed740a69d11201f5ed297c5732f562c6e4e"
+	network         = "prater"
+)
+
+func TestDepositData(t *testing.T) {
+	depositData, err := DepositDataFromFile(t, depositDataFile)
```
Comment (corverroos) [inline eth2util/deposit/deposit_data_test.go:44]: I do not think you need to ever read deposit data from file, just calculate the expected value and do golden file comparison.

**Review by corverroos — CHANGES_REQUESTED** [review]

File: eth2util/deposit/deposit_data.go (hunk 6 above)
Comment (corverroos) [reply eth2util/deposit/deposit_data.go:94]: rename `NewDepositData` to `MarshalDepositData`

File: eth2util/deposit/deposit_data.go [hunk 10]
```diff
@@ -18,142 +18,219 @@ package deposit
...
 
-	// ForkVersion identifies the network/chainID.
-	ForkVersion eth2p0.Version
+	return hashRoot, signingRoot, nil
 }
 
-func (d depositData) HashTreeRoot() ([32]byte, error) {
-	b, err := ssz.HashWithDefaultHasher(d)
+// GetDataRoot returns both the hash root and the signing root of the deposit data.
+func GetDataRoot(pubkey eth2p0.BLSPubKey, withdrawalCreds [32]byte, sig eth2p0.BLSSignature, forkVersion eth2p0.Version) (eth2p0.Root, eth2p0.Root, error) {
```
Comment (corverroos) [inline eth2util/deposit/deposit_data.go:70]: only one of the two responses is ever used, so suggest only retuning that one thing

File: eth2util/deposit/deposit_data.go (hunk 5 above)
Comment (leolara) [reply eth2util/deposit/deposit_data.go:49]: | when returning identical types, it is best to use named parameters so it is clear which one is which.

very good!! let's make code self documenting, we should name return values unless they are super obvious

File: eth2util/deposit/deposit_data.go (hunk 6 above)
Comment (leolara) [reply eth2util/deposit/deposit_data.go:94]: Yes, I agree `New` should be reserved for constructors.

let's as you say unexport as default, and export only things that we need outside

File: eth2util/deposit/deposit_data.go (hunk 7 above)
Comment (xenowits) [reply eth2util/deposit/deposit_data.go:45]: there isn't a proper type for `zeroBytes32`

File: eth2util/deposit/deposit_data.go (hunk 10 above)
Comment (xenowits) [reply eth2util/deposit/deposit_data.go:70]: yes corrected that

File: eth2util/deposit/deposit_data.go (hunk 5 above)
Comment (xenowits) [reply eth2util/deposit/deposit_data.go:49]: changed this to return only one

File: eth2util/deposit/deposit_data_test.go (hunk 9 above)
Comment (xenowits) [reply eth2util/deposit/deposit_data_test.go]: done!

File: eth2util/deposit/deposit.go [hunk 11]
```diff
@@ -13,6 +13,7 @@
 // You should have received a copy of the GNU General Public License along with
 // this program.  If not, see <http://www.gnu.org/licenses/>.
 
+// Package deposit provides functions to create deposit data files.
```
Comment (corverroos) [inline eth2util/deposit/deposit.go:16]: package level godoc is important

File: eth2util/deposit/deposit.go [hunk 12]
```diff
@@ -30,34 +31,27 @@ import (
...
+	eth1AddressWithdrawalPrefix = []byte{0x01}
 
 	// the amount of ether in gwei required to activate a validator.
 	validatorAmt = eth2p0.Gwei(32000000000)
 
-	// zeroes11 refers to a zeroed out 11 byte array.
-	zeroBytes11 = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
-
-	// zeroes refers to a zeroed out 32 byte array.
-	zeroBytes32 = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
```
Comment (corverroos) [inline eth2util/deposit/deposit.go]: zero types not required

File: eth2util/deposit/deposit.go [hunk 13]
```diff
@@ -30,34 +31,27 @@ import (
...
-
 	// DOMAIN_DEPOSIT. See spec: https://benjaminion.xyz/eth2-annotated-spec/phase0/beacon-chain/#domain-types
 	depositDomainType = eth2p0.DomainType([4]byte{0x03, 0x00, 0x00, 0x00})
 )
 
-// GetMessageRoot returns the hash root of the deposit message.
-func GetMessageRoot(pubkey eth2p0.BLSPubKey, withdrawalAddr string) (eth2p0.Root, error) {
+// getMessageRoot returns a deposit message hash root created by the parameters.
+func getMessageRoot(pubkey eth2p0.BLSPubKey, withdrawalAddr string) (eth2p0.Root, error) {
```
Comment (corverroos) [inline eth2util/deposit/deposit.go:44]: make it private

File: eth2util/deposit/deposit.go [hunk 14]
```diff
@@ -30,34 +31,27 @@ import (
...
 	creds, err := withdrawalCredsFromAddr(withdrawalAddr)
 	if err != nil {
-		return eth2p0.Root{}, errors.Wrap(err, "withdrawal credentials")
```
Comment (corverroos) [inline eth2util/deposit/deposit.go]: only wrap external packages, no need to wrap Charon packages or Charon functions (since the errors have proper stack traces)

File: eth2util/deposit/deposit.go [hunk 15]
```diff
@@ -66,101 +60,87 @@ func GetMessageRoot(pubkey eth2p0.BLSPubKey, withdrawalAddr string) (eth2p0.Root
 	return hashRoot, nil
 }
 
-// GetDataRoot returns the hash root of the deposit data.
-func GetDataRoot(pubkey eth2p0.BLSPubKey, withdrawalAddr string, sig eth2p0.BLSSignature) (eth2p0.Root, error) {
```
Comment (corverroos) [inline eth2util/deposit/deposit.go]: this function doesn't do much, so I just inlined it

File: eth2util/deposit/deposit.go [hunk 16]
```diff
@@ -66,101 +60,87 @@ func GetMessageRoot(pubkey eth2p0.BLSPubKey, withdrawalAddr string) (eth2p0.Root
...
-	dmHashRoot, err := GetMessageRoot(pubkey, withdrawalAddr)
+	msgRoot, err := getMessageRoot(pubkey, withdrawalAddr)
 	if err != nil {
-		return nil, errors.Wrap(err, "deposit message root")
+		return nil, err
 	}
 
-	// calculate depositData root
-	ddHashRoot, err := GetDataRoot(pubkey, withdrawalAddr, sig)
+	// TODO(corver): Verify signature matches msgRoot.
```
Comment (corverroos) [inline eth2util/deposit/deposit.go:78]: added this TODO

File: eth2util/deposit/deposit.go [hunk 17]
```diff
@@ -66,101 +60,87 @@ func GetMessageRoot(pubkey eth2p0.BLSPubKey, withdrawalAddr string) (eth2p0.Root
...
 }
 
-// GetDomain returns the Signature Domain.
-func GetDomain(forkVersion eth2p0.Version, domainType eth2p0.DomainType) (eth2p0.Domain, error) {
-	// For deposit, genesisValidatorsRoot is a fixed value.
-	var genesisValidatorsRoot eth2p0.Root
-	copy(genesisValidatorsRoot[:], zeroBytes32)
-
+// getDepositDomain returns the deposit signature domain.
+func getDepositDomain(forkVersion eth2p0.Version) (eth2p0.Domain, error) {
```
Comment (corverroos) [inline eth2util/deposit/deposit.go:109]: more specific function signature, since only used in a single specific way

File: eth2util/deposit/deposit.go [hunk 18]
```diff
@@ -66,101 +60,87 @@ func GetMessageRoot(pubkey eth2p0.BLSPubKey, withdrawalAddr string) (eth2p0.Root
...
 	copy(domain[4:], root[:])
 
 	return domain, nil
 }
 
-// GetSigningRoot returns the signing root by combining a hash root with the deposit domain.
-func GetSigningRoot(forkVersion eth2p0.Version, root eth2p0.Root) ([32]byte, error) {
-	domain, err := GetDomain(forkVersion, depositDomainType)
+// GetMessageSigningRoot returns the deposit message signing root created by the provided parameters.
+func GetMessageSigningRoot(pubkey eth2p0.BLSPubKey, withdrawalAddr string, network string) ([32]byte, error) {
```
Comment (corverroos) [inline eth2util/deposit/deposit.go:127]: this is the new "Step1" function of this package

File: eth2util/deposit/deposit.go [hunk 19]
```diff
@@ -170,35 +150,22 @@ func withdrawalCredsFromAddr(addr string) ([32]byte, error) {
...
 	}
 
-	// Finally, append 20 bytes of ethereum address.
-	creds = append(creds, addrBytes...)
-
-	var resp [32]byte
-	copy(resp[:], creds)
+	var creds [32]byte
+	copy(creds[0:], eth1AddressWithdrawalPrefix) // Add 1 byte prefix.
+	copy(creds[12:], addrBytes)                  // Add 20 bytes of ethereum address suffix.
```
Comment (corverroos) [inline eth2util/deposit/deposit.go:160]: more succinct to work directly with array, then zero11 also not required.

File: eth2util/deposit/deposit.go [hunk 20]
```diff
@@ -207,6 +174,8 @@ func networkToForkVersion(network string) eth2p0.Version {
 		return [4]byte{0x70, 0x00, 0x00, 0x69}
 	case "gnosis":
 		return [4]byte{0x00, 0x00, 0x00, 0x64}
+	case "mainnet": // Default to mainnet
+		fallthrough
 	default:
 		return [4]byte{0x00, 0x00, 0x00, 0x00}
```
Comment (corverroos) [inline eth2util/deposit/deposit.go:180]: it isn't obviously that this is mainnet

File: eth2util/deposit/deposit_test.go [hunk 21]
```diff
@@ -17,53 +17,50 @@ package deposit_test
...
-
 	pubkey, err := tblsconv.KeyToETH2(pk)
 	require.NoError(t, err)
 
-	// Get deposit message root
-	msgRoot, err := deposit.GetMessageRoot(pubkey, withdrawalAddr)
-	require.NoError(t, err)
-
-	forkVersion := eth2p0.Version([4]byte{0x00, 0x00, 0x10, 0x20})
-	msgSigningRoot, err := deposit.GetSigningRoot(forkVersion, msgRoot)
```
Comment (corverroos) [inline eth2util/deposit/deposit_test.go]: now this is just a single step. Also aligned use of network between both functions. Also aligned function signatures

File: eth2util/deposit/deposit_test.go [hunk 22]
```diff
@@ -17,53 +17,50 @@ package deposit_test
...
-
 	sigEth2 := tblsconv.SigToETH2(s)
 
-	// check if serialized versions match.
-	serializedDepositData, err := deposit.MarshalDepositData(pubkey, withdrawalAddr, sigEth2, network)
+	// Check if serialized versions match.
+	actual, err := deposit.MarshalDepositData(pubkey, withdrawalAddr, network, sigEth2)
 	require.NoError(t, err)
 
-	testutil.RequireGoldenBytes(t, serializedDepositData)
```
Comment (corverroos) [inline eth2util/deposit/deposit_test.go]: golden files allow updating, this is not applicable in this usecase

**Review by corverroos — APPROVED** [review]

File: eth2util/deposit/deposit.go (hunk 11 above)
Comment (xenowits) [reply eth2util/deposit/deposit.go:16]: noted!

File: eth2util/deposit/deposit.go (hunk 14 above)
Comment (xenowits) [reply eth2util/deposit/deposit.go]: noted!

File: eth2util/deposit/deposit.go (hunk 17 above)
Comment (xenowits) [reply eth2util/deposit/deposit.go:109]: makes sense👍

File: eth2util/deposit/deposit.go (hunk 19 above)
Comment (xenowits) [reply eth2util/deposit/deposit.go:160]: yes, since the default zero value of a byte is `0`

File: eth2util/deposit/deposit.go (hunk 20 above)
Comment (xenowits) [reply eth2util/deposit/deposit.go:180]: oh i get, this is the first usecase where `fallthrough` makes sense. thanks for pointing this!

File: eth2util/deposit/deposit_test.go (hunk 22 above)
Comment (xenowits) [reply eth2util/deposit/deposit_test.go]: got it. So, for test files like this which are static, golden files are not required👍

Comment (corverroos) [reply eth2util/deposit/deposit_test.go]: yeah, golden files are generated, in this case the test file is a static external test file
